### ISSUE TITLE
If Z3 can't be run, Dafny sometimes hangs instead of reporting an error

### ISSUE THREAD

**Issue by mattmccutchen-amazon** [dafny-lang/dafny#2370]
In some cases in which Dafny can't find Z3, Dafny hangs instead of reporting the expected error "Cannot find any prover executable".  This can make it confusing for the user to troubleshoot the problem.  Failure to find Z3 isn't likely to happen when using a binary release of Dafny that has Z3 bundled at the right path, but it can happen when building Dafny from source or using a custom build system.

**Steps to reproduce** (tested on my Linux machine):
1. Download and extract https://github.com/dafny-lang/dafny/releases/download/v3.7.1/dafny-3.7.1-x64-ubuntu-16.04.zip.
2. Delete the `dafny/z3` directory.
3. Save the example code below to `hang-test.dfy`.
4. Run: `./dafny/dafny /compile:0 hang-test.dfy`

Example code:
```dafny
lemma Dummy1(x: int)
  requires x == 1
  ensures x + 1 == 2 {}

lemma Dummy2(x: int)
  requires x == 1
  ensures x + 1 == 2 {}

lemma Dummy3(x: int)
  requires x == 1
  ensures x + 1 == 2 {}
```

**Expected behavior:** Dafny exits with an error message, "Cannot find any prover executable".

**Actual behavior:** Dafny hangs.

Some further testing I did suggests that the hang happens when the number of methods to be verified is greater than the value of the `/vcsCores` option, which defaults to 1.  With the example code above (which has 3 methods), I observe:
- `./dafny/dafny /compile:0 /vcsCores:3 hang-test.dfy` -> `Cannot find any prover executable` (within a few seconds)
- `./dafny/dafny /compile:0 /vcsCores:2 hang-test.dfy` -> hang

So this may be some kind of deadlock in a thread pool.

**Comment by mattmccutchen-amazon** [comment]
I just noticed #1914.  The two issues may have the same root cause, in which case it probably makes sense to broaden one of them and mark the other as a duplicate.

**Comment by robin-aws** [comment]
> I just noticed #1914. The two issues may have the same root cause, in which case it probably makes sense to broaden one of them and mark the other as a duplicate.

I suspect they will need separate fixes, since the CLI should print out an error but the language server will need to provide a different notification. Perfectly fine for one PR to fix both though. :)

**Comment by keyboardDrummer** [comment]
I've had Dafny verification hang when it passed incorrect parameters to Z3, which happened because it passed parameters for Z3 4.8.5, while it was using Z3 4.8.8

**Comment by MikaelMayer** [comment]
How long ago did you have that issue @keyboardDrummer ?

**Comment by keyboardDrummer** [comment]
Today

**Comment by MikaelMayer** [comment]
CLI or language server? If CLI, what was the command line you ran?

**Comment by keyboardDrummer** [comment]
The language server, it passed `O:model_compress=false` to Z3 4.8.8 which then reports a solver error because it expects `O:model.compact=false`, but Dafny keeps waiting on a Boogie Task that never completes.

**Comment by MikaelMayer** [comment]
Oh that's concerning.
To attach a debugger, I found that you can add Thread.sleep(15000) at the beginning of Program, which gives you 15 seconds to attach the debugger.

**Comment by atomb** [comment]
This still seems to occur sometimes. We should more thoroughly check the case where a z3 binary exists but can't be executed for some reason (being an incompatible version, having missing shared library dependencies, etc.).

**Comment by atomb** [comment]
> Do we currently run any kind of "smoke test" against the specified solver binary, to check that the binary is available and provides a known result, before using it? I see that we run `z3 -version` but fail silently if we don't get anything back ("might be another solver").

Not at the moment. I think it might be good to do that. And I think perhaps we should fail if we don't get anything back from `z3 -version`, since that should always produce something. I think perhaps we should only support other solvers  (e.g., CVC5) if they're explicitly specified with a path to the executable, since using them is very experimental.

**Comment by mattmccutchen-amazon** [comment]
Stepping back: How did a failure to run the solver cause Dafny to hang in the first place?  Can we change that hang to an error rather than relying solely on up-front smoke tests?  No matter how many smoke tests we add, we may overlook some unusual scenario that may affect some user.

**Comment by atomb** [comment]
I've looked into this a little, and it's unfortunately a bit tricky to do. From what I can tell, the .NET `Process` class that we use to invoke the solver hangs if you invoke it with an executable that doesn't exist or can't be executed for some other reason. I would have expected it to throw an exception, but it doesn't seem to. I suspect there's a way to make it work, but it's not as straightforward as it should be.

**Comment by atomb** [comment]
Thanks for digging so deeply into debugging the issue! That should be a straightforward fix to apply.

**Comment by atomb** [comment]
And @keyboardDrummer ended up fixing it: https://github.com/boogie-org/boogie/pull/719

I'll leave this open until we've updated Dafny to depend on that Boogie version and tested it from the Dafny side, though.

**Comment by atomb** [comment]
The Boogie fix was incorporated in 6675361de97be53863b41fd6423dc9d5b5bf001e. The problem doesn't appear to exist anymore, in the tests I've tried.
